src/charm.py

```diff
@@ -1,103 +1,141 @@
 
 import ops
 
-# Log messages can be retrieved using juju debug-log
+import builder
+import cron
+import image
+import openstack_manager
+from state import CharmConfigInvalidError, CharmState
+
 logger = logging.getLogger(__name__)
 
-VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
 
+class GithubRunnerImageBuilderCharm(ops.CharmBase):
+    """Charm the service.
+
+    Attributes:
+        on: Represents custom events managed by cron.
+    """
 
-class IsCharmsTemplateCharm(ops.CharmBase):
-    """Charm the service."""
+    on = cron.CronEvents()
 
-    def __init__(self, *args):
+    def __init__(self, *args: Any):
+        """Initialize the charm.
+
+        Args:
+            args: The CharmBase initialization arguments.
+        """
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
+        self.image_observer = image.Observer(self)
+        self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.build_image_action, self._on_build_image_action)
+        self.framework.observe(self.on.trigger, self._on_cron_trigger)
 
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent):
-        """Define and start a workload using the Pebble API.
+    def _load_state(self) -> CharmState | None:
+        """Load the charm state if valid, set charm to blocked if otherwise.
+
+        Returns:
+            Initialized charm state if valid charm state. None if invalid charm state was found.
+        """
+        try:
+            return CharmState.from_charm(self)
+        except CharmConfigInvalidError as exc:
+            self.unit.status = ops.BlockedStatus(str(exc))
+            return None
 
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
+    def _on_install(self, event: ops.InstallEvent) -> None:
+        """Handle installation of the charm.
 
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
+        Installs apt packages required to build the image.
+
+        Args:
+            event: The event fired on install hook.
+        """
+        self.unit.status = ops.MaintenanceStatus("Setting up Builder.")
+        state = self._load_state()
+        if not state:
+            # Defer this event since on install should be re-triggered to setup dependencies for
+            # the charm. Since the charm goes into blocked state and the user reconfigures the
+            # state, config_changed event should be queued after the deferred on_install.
+            event.defer()
+            return
+
+        builder.configure_proxy(proxy=state.proxy_config)
+        builder.setup_builder()
+        self.unit.status = ops.WaitingStatus("Waiting for first image build.")
```

Comment:
As discussed, this should probably be active since the workload is up and running and busy building the first image

Reply:
Perhaps Waiting for first image could be the status message